File: xabber/src/main/java/com/xabber/android/data/extension/retract/ReplaceMessageIQ.java
package com.xabber.android.data.extension.retract;

import org.jivesoftware.smack.packet.IQ;
import org.jivesoftware.smack.packet.Message;

public class ReplaceMessageIQ extends IQ {
    public static final String NAMESPACE = RetractManager.NAMESPACE;
    public static final String ELEMENT = "replace";
    public static final String ID_ATTRIBUTE = "id";
    public static final String BY_ATTRIBUTE = "by";

    private String id, by;
    private Message message;

    ReplaceMessageIQ(String id, String by, Message message){
        super(ELEMENT, NAMESPACE);
        this.id = id;
        this.by = by;
        this.message = message;
        this.setType(Type.set);
    }

    @Override
    protected IQChildElementXmlStringBuilder getIQChildElementBuilder(IQChildElementXmlStringBuilder xml) {
        xml.attribute(BY_ATTRIBUTE, by);
        xml.attribute(ID_ATTRIBUTE, id);
        xml.rightAngleBracket();
        xml.append(message.toXML());
        return xml;
    }
}

File: xabber/src/main/java/com/xabber/xmpp/retract/outgoing/ReplaceMessageIq.kt
package com.xabber.xmpp.retract.outgoing

import com.xabber.android.data.entity.AccountJid
import com.xabber.android.data.entity.ContactJid
import org.jivesoftware.smack.packet.Message

class ReplaceMessageIq(
    private val messageStanzaId: String,
    private val accountJid: AccountJid,
    private val message: Message,
    archiveAddress: ContactJid? = null
) : AbstractRetractIq(archiveAddress, ELEMENT) {

    init {
        type = Type.set
    }

    override fun getIQChildElementBuilder(xml: IQChildElementXmlStringBuilder) = xml.apply {
        attribute(BY_ATTRIBUTE, accountJid.bareJid.toString())
        attribute(ID_ATTRIBUTE, messageStanzaId)
        rightAngleBracket()
        append(message.toXML())
    }

    private companion object {
        const val ELEMENT = "replace"
        const val BY_ATTRIBUTE = "by"
        const val ID_ATTRIBUTE = "id"
    }

}